kv: generalize lease stasis, accept future-time requests under valid lease

nvb opened the pull request:
Relates to #57688.

This PR performs a series of cleanup steps and eventually generalizes the
replica lease check to consult a request's specific timestamp instead of
assuming that all requests operate at or below `time.Now()`. It then adds to this
generalization some safeguards to ensure that future-time requests don't get
themselves into infinite lease extension/acquisition loops or cause other kinds
of problems due to the requests being too far into the future.

It doesn't go as far as properly handling lease transfers or merged in the case
that a leaseholder has served future-time operations, but that will follow
shortly in another PR.

nvb committed:
kv: generalize STASIS LeaseState to consider future-time reqs

This commit renames the STASIS LeaseState to UNUSABLE and updates
comments to distinguish between the impact that the current time
has on a lease status vs. the impact that a given request's time
has on a lease status. It doesn't yet make any code changes.

nvb committed:
kv: make Lease.Start and LeaseStatus.Now ClockTimestamps

This commit updates the Go type of the `Lease.Start` and `LeaseStatus.Now`
fields to be more strictly typed. Instead of being a `Timestamp` type, they now
are `ClockTimestamp` types, which more effectively ensures that they come from a
real clock. This prevents mistakes and makes their role in the leasing protocol
more clear. For instance, it makes it clear that it is not possible to create a
lease that starts in the future.

This results in a large amount of plumbing, which is all fine.

nvb committed:
kv: restructure lease methods, prepare for request timestamp

This commit restructures the methods in `replica_range_lease.go`
to prepare for time when we will pass a request timestamp into
Replica.leaseStatus. For now, we don't actually pass the request
time in, but we will in a future commit.

nvb committed:
kv: take lease by value in newNotLeaseHolderError

This was causing Lease / LeaseStatus objects to escape to the heap at a few
of its callers, even if the error path was never hit. It was generally a
footgun, so avoid it.

nvb committed:
kv: accept future-time requests under valid lease

This commit generalizes the replica lease check to consult a request's
specific timestamp instead of assuming that all requests operate at or
below time.Now(). It then adds to this generalization some safeguards to
ensure that future-time requests don't get themselves into infinite
lease extension/acquisition loops or cause other kinds of problems due
to the requests being too far into the future.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/58904)
<!-- Reviewable:end -->

nvanbenschoten commented:
This is currently failing CI due to the [last commit](https://github.com/cockroachdb/cockroach/pull/58904/commits/91f58afa5f6063befb2aa1a3fc21ede56d87f66a), but that's not actually meant to be part of this PR. Please ignore that.

nvb committed:
kv/batcheval: document why EvalCtx.GetLease is used over Request.PrevLease

When evaluating RequestLease / TransferLease requests, we compare against the
range's current lease, not the provided PrevLease in the request. This was
confusing and looked like a mistake. I now think it was deliberate, but it
deserved a comment.

tbg reviewed (commented):
> Reviewed 5 of 5 files at r1, 63 of 63 files at r2, 14 of 14 files at r3, 5 of 5 files at r4, 4 of 4 files at r5, 2 of 2 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58904)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/queue.go, line 652 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRPctNm0BEzUl8fNwAi:-MRPctNm0BEzUl8fNwAj:b-b68oat) ([raw file](https://github.com/cockroachdb/cockroach/blob/577e53162c3de3786f65a8b98ec0d0213e2211f8/pkg/kv/kvserver/queue.go#L652)):*
> ```Go
> 		if st.IsValid() && !st.OwnedBy(repl.StoreID()) {
> 			if log.V(1) {
> 				log.Infof(ctx, "needs lease; not adding: %s", st.Lease.String())
> ```

We shouldn't have to call `.String()` here. Does `st.Lease` implement the wrong `fmt.Stringer` on the value type? Can we fix that instead?
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 489 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRPguJw5GCXnUkv9cq5:-MRPguJw5GCXnUkv9cq6:bnmq810) ([raw file](https://github.com/cockroachdb/cockroach/blob/b43772261af1bd515c66b84e4667313fc1faffd6/pkg/kv/kvserver/replica_proposal_buf.go#L489)):*
> ```Go
> 		if !iAmTheLeader && p.Request.IsLeaseRequest() {
> 			// TODO DURING REVIEW: I needed this to deflake a test. But is
> 			// generally seems like a good idea.
> ```

Hmm, yes, it does.
___
*[pkg/kv/kvserver/replica_range_lease.go, line 507 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRPdT-F2e6jv1rtUZaB:-MRPdT-F2e6jv1rtUZaC:b-f7tdfl) ([raw file](https://github.com/cockroachdb/cockroach/blob/577e53162c3de3786f65a8b98ec0d0213e2211f8/pkg/kv/kvserver/replica_range_lease.go#L507)):*
> ```Go
> // its state, unless state == leaseError.
> //
> // - The lease is considered valid if the current timestamp (now) is
> ```

"considered valid" suggests that the returned lease state would be valid. I think what this really tells me is that `VALID` is no longer a good name for the lease state. How about "usable", which indicates the intersection of "not expired at current clock yet" and "can actually serve that particular request"?
___
*[pkg/kv/kvserver/replica_range_lease.go, line 556 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRPdz-z4EWgYJWzqGcR:-MRPdz-z4EWgYJWzqGcS:bo9gkur) ([raw file](https://github.com/cockroachdb/cockroach/blob/577e53162c3de3786f65a8b98ec0d0213e2211f8/pkg/kv/kvserver/replica_range_lease.go#L556)):*
> ```Go
> //   * the client fails to read their own write.
> //
> // - The lease is considered expired in all other cases.
> ```

This misses the proscribed case.
___
*[pkg/kv/kvserver/store_snapshot.go, line 669 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRPey3z8s8Rl3dcathb:-MRPey3z8s8Rl3dcathc:b-hxx9yk) ([raw file](https://github.com/cockroachdb/cockroach/blob/577e53162c3de3786f65a8b98ec0d0213e2211f8/pkg/kv/kvserver/store_snapshot.go#L669)):*
> ```Go
> 					return true
> 				}
> 				// TODO(benesch): this check does detect inactivity on replicas with
> ```

while you're here, add the missing "not"
___
*[pkg/kv/kvserver/kvserverpb/lease_status.proto, line 29 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRPbHRQA__913ptHxVQ:-MRPbHRQA__913ptHxVR:b-hm8z5i) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c5973e4b82f7b8b6990f325d86a5e26d502a7ef/pkg/kv/kvserver/kvserverpb/lease_status.proto#L29)):*
> ```protobuf
>   // UNUSABLE indicates that a lease has not expired at the present time
>   // (i.e. it is VALID), but cannot be used to serve a given request. A
>   // lease may be unusable for one of two reason.
> ```

reasons
___
*[pkg/kv/kvserver/kvserverpb/lease_status.proto, line 70 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRPbX1S378S2QFgQ2o_:-MRPbX1S378S2QFgQ2oa:b-cwy60p) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c5973e4b82f7b8b6990f325d86a5e26d502a7ef/pkg/kv/kvserver/kvserverpb/lease_status.proto#L70)):*
> ```protobuf
>   // PROSCRIBED indicates that the lease's proposed timestamp is earlier
>   // than allowed and can't be used to serve a request. This is used to
>   // detect node restarts: a node that has restarted will see its former
> ```

Isn't it also used in transfers?


<!-- Sent from Reviewable.io -->

andreimatei reviewed (approved):
LGTM

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-:-MRSDwsl0zIjUwb8rfgX:bb6ywmx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @tbg)

___
*[pkg/kv/kvserver/queue.go, line 652 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRPctNm0BEzUl8fNwAi:-MRS1DzZDSLPUkIzVflR:b-hye6wj) ([raw file](https://github.com/cockroachdb/cockroach/blob/577e53162c3de3786f65a8b98ec0d0213e2211f8/pkg/kv/kvserver/queue.go#L652)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

We shouldn't have to call `.String()` here. Does `st.Lease` implement the wrong `fmt.Stringer` on the value type? Can we fix that instead?
</blockquote></details>

I think he might be trying to avoid `st` escaping
___
*[pkg/kv/kvserver/replica_range_lease.go, line 507 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRS4_AF4SIuNU9yRD_M:-MRS4_AF4SIuNU9yRD_N:b57xpfr) ([raw file](https://github.com/cockroachdb/cockroach/blob/577e53162c3de3786f65a8b98ec0d0213e2211f8/pkg/kv/kvserver/replica_range_lease.go#L507)):*
> ```Go
> // its state, unless state == leaseError.
> //
> // - The lease is considered valid if the current timestamp (now) is
> ```

I appreciate the effort to make this comment good.
One thing that I think is not sufficiently clear is the explanation about "now", and why anything about the present time is necessary. One might reasonably ask why the `reqTS` is not sufficient - like it is for sql schema leases, for example. Like, if I'm querying about yesterday, why does the relationship of whatever lease the replica has to the "present time" matter? I could articulate some stuff, but I think you can do it better.
___
*[pkg/kv/kvserver/replica_range_lease.go, line 562 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRS47HZ6d4glbjiFBLs:-MRS47HZ6d4glbjiFBLt:b-vbnmld) ([raw file](https://github.com/cockroachdb/cockroach/blob/577e53162c3de3786f65a8b98ec0d0213e2211f8/pkg/kv/kvserver/replica_range_lease.go#L562)):*
> ```Go
> 	lease roachpb.Lease,
> 	now hlc.ClockTimestamp,
> 	minProposedTS hlc.ClockTimestamp,
> ```

not new, but since you're all up in here: `minProposedTS` is always `r.mu.minProposedTS`. Let's remove the argument.
___
*[pkg/kv/kvserver/replica_range_lease.go, line 613 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRS1mrl0WqPUN_6HCTV:-MRS1mrl0WqPUN_6HCTW:b-7ot3b9) ([raw file](https://github.com/cockroachdb/cockroach/blob/577e53162c3de3786f65a8b98ec0d0213e2211f8/pkg/kv/kvserver/replica_range_lease.go#L613)):*
> ```Go
> //
> // Common operations to perform on the resulting status are to check if
> // is is valid using the IsValid method and to check whether the lease
> ```

is is
___
*[pkg/kv/kvserver/replica_range_lease.go, line 871 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRS8gzXBS2wOlPscvEm:-MRS8gzXBS2wOlPscvEn:b34kqei) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e347b63936eb0d4223b663915ec5b8c5d0e03e0/pkg/kv/kvserver/replica_range_lease.go#L871)):*
> ```Go
> // its output should be completely determined by its parameters.
> func newNotLeaseHolderError(
> 	l roachpb.Lease, proposerStoreID roachpb.StoreID, rangeDesc *roachpb.RangeDescriptor, msg string,
> ```

this makes me sad. Have you tried making a deep copy when assigning `err.Lease`? I'm thinking that should be enough to keep `l` from escaping.
Or, perhaps `LeaseStatus.Lease` should be a pointer. It's nice to make it non-nullable in the proto, but there's no good reason why `LeaseStatus` is even a proto. If it wasn't a proto, then the leases in it would probably always be on the heap anyway - and we might save some copying. No?
___
*[pkg/kv/kvserver/replica_range_lease.go, line 930 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRSDLyX5dKeekRZGKx6:-MRSDLyX5dKeekRZGKx7:bcutray) ([raw file](https://github.com/cockroachdb/cockroach/blob/07119d135f7e28f34af9d5f7e1ce45fcbfa13175/pkg/kv/kvserver/replica_range_lease.go#L930)):*
> ```Go
> }
> 
> // checkRequestTimeRLocked checks that the provided request timestamp is not
> ```

have you done the math about what the timestamp of non-blocking txns will be, and how it relates to the lease renewal policy? Will these requests generally encounter a lease that is good for them?
___
*[pkg/kv/kvserver/replica_range_lease.go, line 942 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRSDBFLDN9RLIRyBmja:-MRSDBFM9f5nP1nYd2m8:bnm9220) ([raw file](https://github.com/cockroachdb/cockroach/blob/07119d135f7e28f34af9d5f7e1ce45fcbfa13175/pkg/kv/kvserver/replica_range_lease.go#L942)):*
> ```Go
> func (r *Replica) checkRequestTimeRLocked(
> 	now hlc.ClockTimestamp, reqTS hlc.Timestamp,
> ) *roachpb.Error {
> ```

consider returning `error` here
___
*[pkg/kv/kvserver/kvserverpb/lease_status.proto, line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRS0OEX82HUyqJKCwnZ:-MRS0OEX82HUyqJKCwn_:bzco6ib) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c5973e4b82f7b8b6990f325d86a5e26d502a7ef/pkg/kv/kvserver/kvserverpb/lease_status.proto#L24)):*
> ```protobuf
>   // ERROR indicates that the lease can't be used or acquired.
>   ERROR = 0;
>   // VALID indicates that the lease is valid at the present time and can
> ```

consider moving away from talking about "present time", and describe these statuses in relationship to an explicit query time - namely the one in `LeaseStatus.Timestamp`.
___
*[pkg/kv/kvserver/kvserverpb/lease_status.proto, line 78 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRS0m1WCq1nTwizf5gV:-MRS0m1WCq1nTwizf5gW:bsjuw1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fcfcece4a3ab04dc17067a7f9fae6c5d31ab1f6/pkg/kv/kvserver/kvserverpb/lease_status.proto#L78)):*
> ```protobuf
> 
> // LeaseStatus holds the lease state, the timestamp at which the state
> // is accurate, the timestamp of the request, the lease, and optionally
> ```

the comment change makes it sound like there's now two different timestamps in here, but there aren't. In the next commit you can put a link to `Replica.leaseStatus` though, and explain something a request's timestamp.


<!-- Sent from Reviewable.io -->

nvb committed:
kv: clarify why we need EXPIRED and UNSUABLE lease statuses

Only expired leases may be acquired by other nodes.

nvb reviewed (commented):
Thanks for the reviews!

One thing I wanted to get both of your opinions on is this `BatchRequest.Timestamp` (and by extension, `BatchResponse.Timestamp`) field. For non-transactional requests, it holds the request's timestamp, which is fine. But for transactional requests, it holds the transaction's read timestamp. This is kind of strange, since as of #32487, we write intents all the way at the transaction's write timestamp. So I think that means two things, independent of this PR:

1. it is possible to write an intent at a timestamp past a leaseholder's current clock
2. it is possible to write an intent at a timestamp past the expiration of a lease

Both of these seem bad, for a few reasons. First, we occasionally bump the timestamp cache to a time based on an intent/value in a range due to server-side retries, and until #59086, it is completely illegal to bump the timestamp cache past a leaseholder's clock. This could result in lost timestamp cache updates on a lease transfer. Second, if a leaseholder can perform intent writes above its local HLC clock then our current proposal to address #36431 is not complete, because even an intent resolved at its write time may violate observed timestamps.

This was all obscured by the fact that lease checks were consulting the local HLC clock, not the `BatchRequest.Timestamp` directly (though the `BatchRequest.Timestamp` was used to update the HLC clock). Now that we're being explicit about the request timestamp and its timestamp, I wonder if there's more we can do here. Maybe we should get rid of this `BatchRequest.Timestamp`/`BatchResponse.Timestamp` field or never set it for transaction requests and then be more explicit about when we want `Txn.ReadTimestamp` (e.g. to pass to `MVCCScan`) vs. when we want `max(Txn.ReadTimestamp, Txn.WriteTimestamp)` (e.g. to compare against a lease)?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-:-MRVux7v0EgISnKZnZt3:bqb0wr0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @tbg)

___
*[pkg/kv/kvserver/queue.go, line 652 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRPctNm0BEzUl8fNwAi:-MRSdd0vBysy5ageOUar:bjj9cdv) ([raw file](https://github.com/cockroachdb/cockroach/blob/577e53162c3de3786f65a8b98ec0d0213e2211f8/pkg/kv/kvserver/queue.go#L652)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think he might be trying to avoid `st` escaping
</blockquote></details>

Right, that was the intention. But even without the `.String()`, it doesn't seem to be escaping. So I removed it.
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 489 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRPguJw5GCXnUkv9cq5:-MRSi5Ji5zMuw4ZLq_Z7:bcv1fjb) ([raw file](https://github.com/cockroachdb/cockroach/blob/b43772261af1bd515c66b84e4667313fc1faffd6/pkg/kv/kvserver/replica_proposal_buf.go#L489)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Hmm, yes, it does.
</blockquote></details>

I opened https://github.com/cockroachdb/cockroach/issues/59179 to track this and removed this commit.
___
*[pkg/kv/kvserver/replica_range_lease.go, line 507 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRS4_AF4SIuNU9yRD_M:-MRSrVdT5JQTa9O-KXOk:bqjezto) ([raw file](https://github.com/cockroachdb/cockroach/blob/577e53162c3de3786f65a8b98ec0d0213e2211f8/pkg/kv/kvserver/replica_range_lease.go#L507)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I appreciate the effort to make this comment good.
One thing that I think is not sufficiently clear is the explanation about "now", and why anything about the present time is necessary. One might reasonably ask why the `reqTS` is not sufficient - like it is for sql schema leases, for example. Like, if I'm querying about yesterday, why does the relationship of whatever lease the replica has to the "present time" matter? I could articulate some stuff, but I think you can do it better.
</blockquote></details>

I've made this change 3 times now before realizing it's unsafe, so that pretty clearly indicates that we need some better comments around here. The reason why we need both a `reqTS` and current clock reading is so that we can differentiate between EXPIRED and UNUSABLE. Only an EXPIRED lease can be non-cooperatively acquired by another node. If a node attempted to grab away an UNUSABLE lease (previously STASIS), it could violate reads served on the existing leaseholder.

I've added some additional assertions and some more documentation to make this clearer. What do you think?
___
*[pkg/kv/kvserver/replica_range_lease.go, line 507 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRPdT-F2e6jv1rtUZaB:-MRVqFvq-UxwJvtjOTFt:bfp2i5v) ([raw file](https://github.com/cockroachdb/cockroach/blob/577e53162c3de3786f65a8b98ec0d0213e2211f8/pkg/kv/kvserver/replica_range_lease.go#L507)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

"considered valid" suggests that the returned lease state would be valid. I think what this really tells me is that `VALID` is no longer a good name for the lease state. How about "usable", which indicates the intersection of "not expired at current clock yet" and "can actually serve that particular request"?
</blockquote></details>

Good point. I reworded a few comments to better indicate this. I didn't get quite as far as renaming valid to usable because I think they both work.
___
*[pkg/kv/kvserver/replica_range_lease.go, line 556 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRPdz-z4EWgYJWzqGcR:-MRSj4UXEqrWE6rfkxVA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/577e53162c3de3786f65a8b98ec0d0213e2211f8/pkg/kv/kvserver/replica_range_lease.go#L556)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This misses the proscribed case.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/replica_range_lease.go, line 562 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRS47HZ6d4glbjiFBLs:-MRVqYiO4Blg9JdbRPvX:b9wqnqn) ([raw file](https://github.com/cockroachdb/cockroach/blob/577e53162c3de3786f65a8b98ec0d0213e2211f8/pkg/kv/kvserver/replica_range_lease.go#L562)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

not new, but since you're all up in here: `minProposedTS` is always `r.mu.minProposedTS`. Let's remove the argument.
</blockquote></details>

Eh, I think the idea is that the status is fully resolved by these arguments, so there are no hidden dependencies. That's also why we accept the lease. But that's also not quite true, because liveness is a dependency. Regardless, I don't feel strongly about this but will save additional refactoring for later.
___
*[pkg/kv/kvserver/replica_range_lease.go, line 613 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRS1mrl0WqPUN_6HCTV:-MRSl1Fj6LKVgs7yjoJM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/577e53162c3de3786f65a8b98ec0d0213e2211f8/pkg/kv/kvserver/replica_range_lease.go#L613)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

is is
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/replica_range_lease.go, line 871 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRS8gzXBS2wOlPscvEm:-MRVqw9w-QH29rAXUCvh:b-uztt4y) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e347b63936eb0d4223b663915ec5b8c5d0e03e0/pkg/kv/kvserver/replica_range_lease.go#L871)):*
> Or, perhaps LeaseStatus.Lease should be a pointer. It's nice to make it non-nullable in the proto, but there's no good reason why LeaseStatus is even a proto. If it wasn't a proto, then the leases in it would probably always be on the heap anyway - and we might save some copying. No?

I looked into this and I think the issue is that we don't have any guarantee that the reference (`r.mu.state.Lease`) won't change once we drop the replica mutex. In practice, I think this would be safe because we only ever replace the reference wholesale. But I don't want to pull on that while making the rest of this change.
___
*[pkg/kv/kvserver/replica_range_lease.go, line 930 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRSDLyX5dKeekRZGKx6:-MRSl41Y60komW5Ey1N_:bd4kdap) ([raw file](https://github.com/cockroachdb/cockroach/blob/07119d135f7e28f34af9d5f7e1ce45fcbfa13175/pkg/kv/kvserver/replica_range_lease.go#L930)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

have you done the math about what the timestamp of non-blocking txns will be, and how it relates to the lease renewal policy? Will these requests generally encounter a lease that is good for them?
</blockquote></details>

Non-blocking txns will never need to write/commit more than 1s in the future. Leases are 9 seconds long and renew every 4.5 seconds, as does node liveness. So we should be good here.
___
*[pkg/kv/kvserver/replica_range_lease.go, line 942 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRSDBFLDN9RLIRyBmja:-MRSlIF50W8F8mPUqDzC:b-nnnrmo) ([raw file](https://github.com/cockroachdb/cockroach/blob/07119d135f7e28f34af9d5f7e1ce45fcbfa13175/pkg/kv/kvserver/replica_range_lease.go#L942)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

consider returning `error` here
</blockquote></details>

This change is made in https://github.com/cockroachdb/cockroach/pull/59086. I'll leave this as-is for now.
___
*[pkg/kv/kvserver/store_snapshot.go, line 669 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRPey3z8s8Rl3dcathb:-MRSlUWZ5ExuXFLATpsa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/577e53162c3de3786f65a8b98ec0d0213e2211f8/pkg/kv/kvserver/store_snapshot.go#L669)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

while you're here, add the missing "not"
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/kvserverpb/lease_status.proto, line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRS0OEX82HUyqJKCwnZ:-MRVqPbv9lL9enBR0anL:bj7apw7) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c5973e4b82f7b8b6990f325d86a5e26d502a7ef/pkg/kv/kvserver/kvserverpb/lease_status.proto#L24)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

consider moving away from talking about "present time", and describe these statuses in relationship to an explicit query time - namely the one in `LeaseStatus.Timestamp`.
</blockquote></details>

See discussion above.
___
*[pkg/kv/kvserver/kvserverpb/lease_status.proto, line 29 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRPbHRQA__913ptHxVQ:-MRSpQcTC87PrITVlA2L:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c5973e4b82f7b8b6990f325d86a5e26d502a7ef/pkg/kv/kvserver/kvserverpb/lease_status.proto#L29)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

reasons
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/kvserverpb/lease_status.proto, line 70 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRPbX1S378S2QFgQ2o_:-MRSpOLN-xIj3YcssBpX:b60r506) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c5973e4b82f7b8b6990f325d86a5e26d502a7ef/pkg/kv/kvserver/kvserverpb/lease_status.proto#L70)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Isn't it also used in transfers?
</blockquote></details>

Yes, done.
___
*[pkg/kv/kvserver/kvserverpb/lease_status.proto, line 78 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRS0m1WCq1nTwizf5gV:-MRSqlr9448Zyvi9ovLV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fcfcece4a3ab04dc17067a7f9fae6c5d31ab1f6/pkg/kv/kvserver/kvserverpb/lease_status.proto#L78)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

the comment change makes it sound like there's now two different timestamps in here, but there aren't. In the next commit you can put a link to `Replica.leaseStatus` though, and explain something a request's timestamp.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
Yeah, we should clean this up. It's likely too hard to remove the `.Timestamp` field completely (what about all of those non-transactional requests like QueryTxn, Increment, etc?) in the short term, but at the very least we should be forwarding it to the write timestamp to avoid the issues you pointed out.
It's always bothered me that too many timestamps were floating around in the lower levels of KV - txn read timestamp, txn write timestamp, batch.Timestamp - perhaps we can avoid any direct access to these timestamps and instead use `.ReadTimestamp()`, `.WriteTimestamp()` getters that can coalesce from these three sources. Not sure, I haven't looked at these code paths in a while.

> Reviewed 73 of 73 files at r8, 63 of 63 files at r9, 14 of 14 files at r10, 5 of 5 files at r11, 4 of 4 files at r12, 2 of 2 files at r13, 3 of 3 files at r14.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-:-MS7jKMw1odp0nmsNlep:b-wrtgjl)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/replica_range_lease.go, line 562 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MRS47HZ6d4glbjiFBLs:-MS7bBxQEv2sArZE_V_O:b7cdvau) ([raw file](https://github.com/cockroachdb/cockroach/blob/577e53162c3de3786f65a8b98ec0d0213e2211f8/pkg/kv/kvserver/replica_range_lease.go#L562)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Eh, I think the idea is that the status is fully resolved by these arguments, so there are no hidden dependencies. That's also why we accept the lease. But that's also not quite true, because liveness is a dependency. Regardless, I don't feel strongly about this but will save additional refactoring for later.
</blockquote></details>

I'd leave as is, if anything this method should eventually become standalone rather than more tightly coupled with the Replica behemoth.
___
*[pkg/kv/kvserver/replica_range_lease.go, line 529 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MS7cKf31ed6JUWasJzx:-MS7cKf31ed6JUWasJzy:b-zhu620) ([raw file](https://github.com/cockroachdb/cockroach/blob/16b3579e86d04faec327b3fc863511bec00275ed/pkg/kv/kvserver/replica_range_lease.go#L529)):*
> ```Go
> // - The lease is considered valid if the current timestamp (now) is
> //   covered by the supplied lease and the lease can be used to serve a
> //   request at the specified time. This is determined differently
> ```

I'm still not convinced by the structure of this comment and the notion of "validity". There are really two questions we ask about leases:

1. can I safely request a non-cooperative lease following the current lease? This is linked to the (lease or epoch) expiration and the `now` timestamp.
2. can the lease serve an operation at timestamp T? The answer is no if the lease has been relinquished at a timestamp >= T (minProposedTS, i.e. lease transfer or a restart), or if T >= stasis.

You didn't bite at my suggestion to rename VALID, but can I coax you into consistently upper-casing the enum variants in the comments to avoid confusion?

Strawman comment rewrite from my input above:

The lease status is linked to the desire to serve a request at a specific timestamp (which may be a future timestamp) under the lease, as well as a notion of the current hlc time (`now`).

A status of ERROR indicates a failure to determine the correct lease status, and should not occur under normal operations. The caller's only recourse is to give up or to retry.

If the lease is expired according to the `now` timestamp (and, in the case of epoch-based leases, the epoch), a status of EXPIRED is returned. Note that this ignores the timestamp of the request, which may well technically be eligible to be served under the lease. The key feature of an EXPIRED status is that it reflects that a new lease with a start timestamp greater than or equal to `now` can be acquired non-cooperatively.

If the lease is not EXPIRED, the request timestamp is taken into account. The expiration timestamp is adjusted for clock offset; if the request timestamp falls into the so-called stasis period at the end of the lifetime of the lease, the status is UNUSABLE, which callers typically want to react to by extending the lease, if they are in a position to do so.

Finally, for request timestamps falling before the stasis period, the request timestamp is additionally checked against the minProposedTimestamp. This timestamp determines the earliest timestamp the lease can be used for, and is set both in cooperative lease transfers and to prevent reuse of leases across node restarts (which would result in latching violations). Requests preceding this timestamp are assigned a status of PROSCRIBED and should be retried at a higher timestamp, or the lease transfer target, depending on the scenario.

On the surface, ...
___
*[pkg/kv/kvserver/kvserverpb/lease_status.proto, line 81 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MS7iMiCEL2kAx2GVTyh:-MS7iMiCEL2kAx2GVTyi:b-khjxvf) ([raw file](https://github.com/cockroachdb/cockroach/blob/16b3579e86d04faec327b3fc863511bec00275ed/pkg/kv/kvserver/kvserverpb/lease_status.proto#L81)):*
> ```protobuf
> 
> // LeaseStatus holds the lease state, the current clock time at which the
> // state is accurate, the request time at which the stats is accurate, the
> ```

stats?


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Yeah, we should clean this up. ...

Sounds like a plan. I'll address this in a follow-up PR.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-:-MSAJYa0CRDRemv9_I4Z:bwdshhf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @tbg)

___
*[pkg/kv/kvserver/replica_range_lease.go, line 529 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MS7cKf31ed6JUWasJzx:-MSAFA3o4csQFGv48tmn:b-v0j1zn) ([raw file](https://github.com/cockroachdb/cockroach/blob/16b3579e86d04faec327b3fc863511bec00275ed/pkg/kv/kvserver/replica_range_lease.go#L529)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I'm still not convinced by the structure of this comment and the notion of "validity". There are really two questions we ask about leases:

1. can I safely request a non-cooperative lease following the current lease? This is linked to the (lease or epoch) expiration and the `now` timestamp.
2. can the lease serve an operation at timestamp T? The answer is no if the lease has been relinquished at a timestamp >= T (minProposedTS, i.e. lease transfer or a restart), or if T >= stasis.

You didn't bite at my suggestion to rename VALID, but can I coax you into consistently upper-casing the enum variants in the comments to avoid confusion?

Strawman comment rewrite from my input above:

The lease status is linked to the desire to serve a request at a specific timestamp (which may be a future timestamp) under the lease, as well as a notion of the current hlc time (`now`).

A status of ERROR indicates a failure to determine the correct lease status, and should not occur under normal operations. The caller's only recourse is to give up or to retry.

If the lease is expired according to the `now` timestamp (and, in the case of epoch-based leases, the epoch), a status of EXPIRED is returned. Note that this ignores the timestamp of the request, which may well technically be eligible to be served under the lease. The key feature of an EXPIRED status is that it reflects that a new lease with a start timestamp greater than or equal to `now` can be acquired non-cooperatively.

If the lease is not EXPIRED, the request timestamp is taken into account. The expiration timestamp is adjusted for clock offset; if the request timestamp falls into the so-called stasis period at the end of the lifetime of the lease, the status is UNUSABLE, which callers typically want to react to by extending the lease, if they are in a position to do so.

Finally, for request timestamps falling before the stasis period, the request timestamp is additionally checked against the minProposedTimestamp. This timestamp determines the earliest timestamp the lease can be used for, and is set both in cooperative lease transfers and to prevent reuse of leases across node restarts (which would result in latching violations). Requests preceding this timestamp are assigned a status of PROSCRIBED and should be retried at a higher timestamp, or the lease transfer target, depending on the scenario.

On the surface, ...

</blockquote></details>

Done. Thanks for the strawman :smile: I just made minor modifications.
___
*[pkg/kv/kvserver/kvserverpb/lease_status.proto, line 81 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/58904#-MS7iMiCEL2kAx2GVTyh:-MSA3XuZ7oFbLi0xUl7_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/16b3579e86d04faec327b3fc863511bec00275ed/pkg/kv/kvserver/kvserverpb/lease_status.proto#L81)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

stats?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
bors r+
